python/lsst/pipe/tasks/findSatellites.py

```diff
@@ -0,0 +1,672 @@
+        if len(self.lines) == 0:
+            lineMask = np.zeros_like(maskData, dtype=bool)
+            fitLines = np.empty(0, dtype=[('rho', float), ('theta', float), ('votes', int)])
+            clusters = np.empty(0, dtype=[('rho', float), ('theta', float), ('votes', int)])
+        else:
+            clusters = self.findClusters(self.lines)
+            fitLines, lineMask = self.fitProfile(clusters, maskedImage)
+
+        # The output mask is the intersection of the fit satellite trails and the image detections
+        outputMask = lineMask & maskData.astype(bool)
+
+        return pipeBase.Struct(
+            lines=fitLines,
+            lineClusters=clusters,
+            originalLines=self.lines,
+            mask=outputMask,
+        )
+
+    def processMaskedImage(self, maskedImage, checkMethod=False):
+        """Make binary image array from maskedImage object
+
+        Convert input to a binary image by setting all data with signal-to-noise
+        below some threshold to zero, and all data above the threshold to one.
+        If the imageBinning config parameter has been set, this procedure will
+        be preceded by a weighted binning of the data.
+
+        Parameters
+        ----------
+        maskedImage : `lsst.afw.image.Exposure`
+            Image to be (optionally) binned and converted
+
+        Returns
+        -------
+        out_data : `np.ndarray`
+            2-d binary image of pixels above the signal-to-noise threshold.
+        """
+        binning = self.config.imageBinning
+
+        data = maskedImage.image.array
+        weights = 1 / maskedImage.variance.array
+        mask = maskedImage.getMask()
+
+        detectionMask = ((mask.array & mask.getPlaneBitMask("DETECTED")))
+        badMask = ((mask.array & mask.getPlaneBitMask("NO_DATA")) |
+                   (mask.array & mask.getPlaneBitMask("INTRP")) |
+                   (mask.array & mask.getPlaneBitMask("BAD")) |
+                   (mask.array & mask.getPlaneBitMask("SAT")) |
+                   (mask.array & mask.getPlaneBitMask("EDGE"))).astype(bool)
+
+        fitMask = detectionMask.astype(bool) & ~(badMask.astype(bool))
+        fitData = np.copy(data)
+        fitData[~fitMask] = 0
+        fitWeights = np.copy(weights)
+        fitWeights[~fitMask] = 0
+
+        if binning is not None:
+            # Do weighted binning:
+            ymax, xmax = fitData.shape
+            if (ymax % binning == 0) & (xmax % binning == 0) & (not checkMethod):
+                # Faster binning method
+                binNumeratorReshape = (fitData * fitWeights).reshape(ymax // binning, binning,
+                                                                     xmax // binning, binning)
+                binDenominatorReshape = fitWeights.reshape(binNumeratorReshape.shape)
+                binnedNumerator = binNumeratorReshape.sum(axis=3).sum(axis=1)
+                binnedDenominator = binDenominatorReshape.sum(axis=3).sum(axis=1)
+            else:
```

Comment:
The algorithm is certainly not ideal. This is doing a binned average, where the last bin will not have as many pixels, so the variance ends up being a bit different from the other bins. However, it has such a small effect overall, that I decided it was okay. We could just have this part result in an error if the user chooses a bin size that doesn't evenly divide the image, but since I don't think it could effect the results, I didn't think that this should be a fatal error.